numpy/_core/src/umath/_scaled_float_dtype.c

```diff
@@ -454,7 +454,7 @@ sfloat_init_casts(void)
         .nout = 1,
         /* minimal guaranteed casting */
         .casting = NPY_SAME_KIND_CASTING,
-        .flags = NPY_METH_SUPPORTS_UNALIGNED,
+        .flags = NPY_METH_SUPPORTS_UNALIGNED | NPY_METH_REQUIRES_PYAPI,
```

Comment:
```suggestion
        .flags = NPY_METH_SUPPORTS_UNALIGNED,
```
If this change is necessary, that is a bug in the error check function, and it is almost surprising more isn't failing (probably defensive `PyErr_Occurred()` checks due to legacy paths).

Reply:
No it's not a bug, the loop uses:
```
    npy_gil_error(PyExc_TypeError,
                  "error raised inside the core-loop: non-finite factor!");
```
to set the error which it is allowed to do and which is a common pattern.  The only reason you don't see this pattern a lot more, is that ufuncs have an additional `PyErr_Occurred()` check that always happens (due to old loops not indicating an error).